Consolidate EuiText and EuiMarkdownFormat styling

miukimiu opened the pull request:
# Summary

This PR is a first step in making the **EuiMarkdownFormat** use the **EuiText**. So the **EuiText** improvements were thought in a way to benefit the **EuiMarkdownFormat** usage.

So given the complexity of this PR, we should improve issues like bottom margins of EUI components inside the **EuiText** in a different PR.  

Or we should open a feature branch just to tackle styling issues that exist in the **EuiText** or in components when they live inside the **EuiText**.  But I would say that this PR closes #4145.

# Render markdown files in our docs

* Probably it's not a necessary feature, but I started this attempt on https://github.com/elastic/eui/pull/4449. There were still a few issues like the "scroll to" functionality not going to the exact position that it's now fixed.
* The MD files are rendered with the **EuiMarkdownFormat**. Each h2 inside the MD files is transformed into a new section/side nav item. 

<img width="1443" alt="Screenshot 2021-07-15 at 18 03 44" src="https://user-images.githubusercontent.com/2750668/125828168-8cf15beb-d8c8-4830-83bf-57c1198ef0f1.png">


# More components with Remark

* The **EuiCodeblock** was fixed to have a better wrapper.
* Blockquotes now uses a specific markdown format class
* Tables now use a specific markdown format class
* HRs are now using the **EuiHorizontalRule** 

# Scale text mixin

The `euiScaleText` mixing had some changes:

* It now accepts a `$sizingMethod` that can be `em` or `rem`. 
* All the font sizes are calculated using a function `fontSizeToRemOrEm($size, $sizingMethod: 'rem')`.   
* All the margin and paddings are calculated using a function called `marginToRemOrEm($elementSize, $elementFontSize, $sizingMethod: 'rem')`.
* The `line-height` used to return `rem` now returns unitless numbers. This way, it behaves the same for both `rem` and `em` font-size scenarios.

# EuiText

* It now receives a new size `relative`. This works for scenarios where we want to scale the font.
* We can also pass a non-named color like a hex value. 
* Blockquotes font size now matches the base font size which is the same as paragraphs.

#  EuiMarkdownFormat styles

Text, Borders, and backgrounds change according to the color passed. They can also inherit or receive any EUI named color.
 With this approach, I don't see the reason for having a `reverseColorsFromTheme` or a `colorMethod` that accepts solid or translucent as an option. 
The decision of the color passed is on the consumer implementation. By knowing the theme context, consumers can pass colors that make sense. Also, the `color` prop accepts `rgba` colors for translucent scenarios.

# Pages to test the new functionalities:

 - https://eui.elastic.co/pr_4663/#/guidelines/getting-started
 - https://eui.elastic.co/pr_4663/#/editors-syntax/markdown-format#kitchen-sink
 - https://eui.elastic.co/pr_4663/#/display/text

### Checklist

- [x] Check against **all themes** for compatibility in both light and dark modes
- ~[ ] Checked in **mobile**~
- [x] Checked in **Chrome**, **Safari**, **Edge**, and **Firefox**
- [x] Props have proper **autodocs** and **[playground toggles](https://github.com/elastic/eui/blob/master/wiki/documentation-guidelines.md#adding-playground-toggles)**
- [x] Added **[documentation](https://github.com/elastic/eui/blob/master/wiki/documentation-guidelines.md)**
- ~[ ] Checked **[Code Sandbox](https://codesandbox.io/)** works for the any docs examples~
- [x] Added or updated **[jest tests](https://github.com/elastic/eui/blob/master/wiki/testing.md)**
- ~[ ] Checked for **breaking changes** and labeled appropriately~
- ~[ ] Checked for **accessibility** including keyboard-only and screenreader modes~
- [x] A **[changelog](https://github.com/elastic/eui/blob/master/wiki/documentation-guidelines.md#changelog)** entry exists and is marked appropriately

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4663/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4663/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4663/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4663/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4663/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4663/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4663/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4663/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4663/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4663/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4663/

cchaos commented on src-docs/src/views/markdown_editor/markdown_format_sink.js:
```diff
@@ -146,5 +159,137 @@ Autoconverted link https://github.com/nodeca/pica (enable linkify to see)
+
+      <EuiSpacer size="s" />
+
+      <EuiPanel hasBorder={true} className={panelClasses}>
```
Can you reduce the panel-within-panel effect this is having?

cchaos commented on src-docs/src/views/markdown_editor/markdown_format_sink.js:
```diff
@@ -146,5 +159,137 @@ Autoconverted link https://github.com/nodeca/pica (enable linkify to see)
+  const closePopover = () => setIsPopoverOpen(false);
+
+  const button = (
+    <EuiButtonEmpty
+      onClick={onButtonClick}
+      iconType="controlsVertical"
+      aria-label="Open props customization"
+      color="text"
+      size="s">
+      Customize props
+    </EuiButtonEmpty>
```
This isn't really standing out as a control, I'd just make it a regular primary button.
```suggestion
    <EuiButton
      onClick={onButtonClick}
      iconType="controlsVertical"
      size="s">
      Customize props
    </EuiButton>
```

cchaos commented on src-docs/src/views/markdown_editor/markdown_format_sink.js:
```diff
@@ -146,5 +159,137 @@ Autoconverted link https://github.com/nodeca/pica (enable linkify to see)
+              <EuiRange
+                min={12}
+                max={24}
+                step={4}
```
When I saw this slider, I didn't expect it to be restricted to certain steps. If you want to keep it restrictive, I'd suggest showing the steps/labels. Otherwise, just remove this restriction and allow any integer.

cchaos commented on src-docs/src/views/markdown_editor/mardown_format_example.js:
```diff
@@ -69,6 +75,32 @@ export const MarkdownFormatExample = {
+      title: 'Text sizing and coloring',
+      text: (
+        <p>
+          <strong>EuiMarkdownFormat</strong> uses the{' '}
```
```suggestion
          <strong>EuiMarkdownFormat</strong> uses{' '}
```

cchaos commented on src-docs/src/views/markdown_editor/mardown_format_example.js:
```diff
@@ -69,6 +75,32 @@ export const MarkdownFormatExample = {
+          code: markdownFormatStylesSource,
+        },
+        {
+          type: GuideSectionTypes.HTML,
```
You don't need to supply the HTML config anymore since we're no longer displaying the HTML tabs. 😉

cchaos commented on src-docs/src/routes.js:
```diff
@@ -293,34 +293,33 @@ const createExample = (example, customTitle) => {
+    name: title,
+    component: () => (
+      <GuidePage title={title}>
+        <GuideMarkdownFormat title={title}>
```
```suggestion
        <GuideMarkdownFormat grow={false} title={title}>
```

I'm a stickler for readable line-lengths. Since this page is just a bulk of text, let's ensure the line-lengths are good.

cchaos commented on src-docs/src/components/guide_page/_guide_page.scss:
```diff
@@ -82,3 +82,10 @@ $guideSideNavWidth: 240px;
+
+// pages using a markdown format
+.guideMarkdownFormat  {
+  h2 {
+    padding-top: $euiSizeL;
+  }
```
I don't think this is necessary anymore.

cchaos reviewed (changes requested):
❤️ I know this was a lot of work, especially the relative sizing and custom coloring. I see a lot of duplicated mixin/functions but I'm not to worried about "DRY" code at the moment because I'm really hoping Emotion will make this easier in the near future.

There are definitely a few changes to the margins/paddings/line-heights of certain text elements. It would be great to track these in a comment so that we can show the intentionality of the changes and to know exactly what has changed. 

---

One specific formatting thing that I'm seeing with the Markdown formatter, is that is only seems to be wrapping multi-line code blocks with our actual EuiCodeBlock (not inline). It would be great that even just single line code blocks could be wrapped in EuiCodeBlock.
<img width="512" alt="Screen Shot 2021-07-19 at 12 13 51 PM" src="https://user-images.githubusercontent.com/549577/126192927-d186eddc-8f04-4022-af16-c41535afd886.png">

snide commented on src/components/markdown_editor/plugins/markdown_default_plugins.tsx:
```diff
@@ -84,6 +85,19 @@ export const getDefaultEuiMarkdownProcessingPlugins = (): [
+        blockquote: (props) => (
+          <blockquote {...props} className="euiMarkdownFormat__blockquote" />
+        ),
+        table: (props) => (
+          <table className="euiMarkdownFormat__table" {...props} />
+        ),
```
Didn't look to deep into this, but figured I'd share since they use similar systems. An alternative way I did this in the docs. Using the full EUI table components means that you get all the nice stuff like the responsive defaults rather than needing to write all the CSS. This example also subs the images out as well, which might be worthwhile.

https://github.com/elastic/elastic-docs-next/blob/master/src/components/mdx/provider_components.tsx#L74-L78

thompsongl commented on src/components/markdown_editor/plugins/markdown_default_plugins.tsx:
```diff
@@ -84,6 +85,19 @@ export const getDefaultEuiMarkdownProcessingPlugins = (): [
           ) : (
             <EuiCode {...props} />
           ),
+        // When we use block code "fences" the code tag is replaced by the `EuiCodeBlock`.
```
I don't expect any trouble, but for visibility, #4970 also [edited this file](https://github.com/elastic/eui/pull/4970/files#diff-4af5cc21c5437d72bf6519e98a616a003248c8c9e11f6e06db8d6c7ba740ce50R58-R75) and will likely merge first.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4663/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4663/

miukimiu commented:
@cchaos I went through your comments and I fixed all of them expect the following one:

> One specific formatting thing that I'm seeing with the Markdown formatter, is that is only seems to be wrapping multi-line code blocks with our actual EuiCodeBlock (not inline). It would be great that even just single line code blocks could be wrapped in EuiCodeBlock.

Right now this part of the code checks if there are linebreaks to use code block or otherwise inline code. 

https://github.com/elastic/eui/blob/004ac1210a12788f58692f0bd8518e31374e8b7d/src/components/markdown_editor/plugins/markdown_default_plugins/processing_plugins.tsx#L64-L70 

I can use the technique mentioned here: https://github.com/rehypejs/rehype-react/issues/6#issuecomment-323868265. If the props contain a `className` like `className='language-js'` is a codeblock, otherwise inline code.

```
code: (props) => {
          return props.className ? (
            <EuiCodeBlock fontSize="m" paddingSize="s" isCopyable {...props} />
          ) : (
            <EuiCode {...props} />
          );
        },
```

But I tried and if consumers don't pass a language like **```js** and just pass triple backticks **```** no `className` is created and with this technique, it assumes it's an inline code and used the <EuiCode {...props} />. So consumers would always have to specify a language when using triple backticks.

So not sure how can we better solve this issue. Ideally, if it's passed triple backticks with or without a language it should use a **EuiCodeBlock** otherwise **EuiCode**.

But I couldn't find better solutions. @thompsongl any idea on how we can solve this?

cchaos commented:
Does that mean that `rehype-react` doesn't support differing between triple backticks and single backticks?

miukimiu commented:
I think the problem starts before. When the code gets to `rehype-react` the node it's already transformed without many options to distinguish if it is inline code or not. 

This seems a good solution: https://github.com/rehypejs/rehype-react/issues/6#issuecomment-323714520. Somehow we need to inject a class property at the node in order to distinguish inline/fenced code. So when it gets to `rehype-react` we know it's inline or not.

thompsongl commented:
Latest commit fixes the inline (single tick) vs. fenced (triple tick) code block discrepancy.
In reality, it had more to do with whether `language` was provided for a fenced block. Regardless, we now get the appropriate code type making the distinction clear in `remark-prismjs`.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4663/

cchaos reviewed (approved):
💯 LGTM! I still think it's a good idea to track the differences in the text element renders. [For instance](https://share.getcloudapp.com/yAur5Zd6) the `blockquote` text size has been reduced. (I think it looks better, but others might not realize).

Maybe you can also update the [home_view link to the new Getting Started page](https://share.getcloudapp.com/JruDr1P2)? 😉 

I'd like an engineer to do a quick pass and then don't forget to merge with upstream to get changelog changes.

snide commented:
cc @gjones You'll want to watch for this one coming downstream after it merges. We utilize the `EuiText` side a bit in docs.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4663/

miukimiu commented:
@cchaos I confirmed if all the margins, paddings, line-heights and font-sizes were matching the old sizes: https://github.com/elastic/eui/blob/master/src/components/text/_text.scss

It seems that I only changed blockquotes to match the base font-size/line-height. I added a CL entry with this change and updated the PR summary.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4663/

thompsongl reviewed (approved):
Code changes LGTM! Thanks for all the work on this 🙏

miukimiu commented:
Thanks @thompsongl and @cchaos for the help! 🎉
